### PR TITLE
Update stacking scripts

### PR DESCRIPTION
**Pull request by Morlic-fo**

### CODE REVIEW

**Comment by MatGB** [comment]
Excellent, thank you, that now makes sense looking at the code, I had missed that the stealth parts also had a capacity now.

File: default/ship_parts.txt [hunk 1]
```diff
@@ -1134,11 +1134,22 @@ Count condition = And [
             ]
 '''
 
+// Beginning of Stacking macros
+/* For the AI and general uniformity, shields, detectors and cloaks are assigned a capacity.
+However, to prevent stacking, for each group X there is a X_STACKING macro which basically 
+first removes all effects of the used parts (by subtracting the summed capacity of all used parts)
```
Comment (geoffthemedio) [inline default/ship_parts.txt:1140]: this ridiculous workaround is why I think that capacity-based effects should not be automatically added to parts, or at least there should be a way to disable them in the part script.

Comment (Vezzra) [inline default/ship_parts.txt:1140]: I'd say this entire thing of complicated scripting to get reasonable stacking behavior working shows that the stacking code on the C++ side needs an overhaul. There needs to be a way to specify in the scripts more precisely how effectsgroups should stack.

The most simple approach would be to be able to specify a "stacking priority" alongside the stacking group in an effectsgroup. The effectsgroup with the highest stacking priority within a stacking group will be applied.

Comment (MatGB) [inline default/ship_parts.txt:1140]: Aye, I think this came about after I accidentally designed a scout with two different scanning parts and got the worst of the two only, the real affect of the workaround was to make non-stacking parts work as expected and not semi-randomly. That it also makes use of capacity to assist the AI and the Pedia displays as well as let us script species traits more easily is a bonus (of sorts).

I would be very happy of the backend was rewritten to negate the need for workarounds but still let us have various parts and pick which should stack (eg Multi Spectral Shields and Trans Spatial Drive both give a stealth bonus, this lets us pick if they should stack or not whereas before we didn't have a huge amount of control without special exceptions in each part).

**Comment by geoffthemedio** [comment]
I think that trying to implement stacking mechanics in C++ at all is a bad idea.

The problem I see here is with the autogenerated effects that need to be undone in order to get the desired stacking behaviour. There should instead at least be a way to disable them (since there's so much resistance to removing them entirely as I would prefer).

For stacking, assuming a particular behaviour in C++ is, I think, unnecessary and probably unproductive. Different parts will need different styles of stacking limitations, and this can most easily be written into the scripts.

**Comment by Vezzra** [comment]
Sorry, but I strongly disagree on that point. First of all, we already _have_ a stacking mechanism implemented in C++, it's just done in a way that makes it very ackward to use. So I wonder what you actually want to propose here - that even the very limited support for stacking we have now should be removed and try to implement all stacking behavior with the means now available in FOCS?

_I_ think that _this_ would be a bad idea. I only skimmed over Morlics code, and in this particular case the autogenerated effects may complicate things even _more_, but even without them implementing stacking exclusively with FOCS _already is_ a needlessly complicated affair.

Adding the possibility to specify a "stacking priority" together with the stacking group doesn't sound like it would require very extensive/complicated coding on the C++ side, but will go a **_long**_ way to make implementing stacking in FOCS much, much easier for scripters. It might not cover all possible stacking behaviors scripters might desire, but it will certainly cover the most common cases and help reduce a lot of complicated scripting.

I really can't follow how such a feature provided by the C++ side would be "unnecessary and probably unproductive". On the contrary, _not_ having it seems much more "unproductive" to me.

P.S.: Of course, adding a possibility to at least disable autogenerated effects sounds like a very reasonable idea too. There probably will be enough cases where such a feature will help avoid complicated scripting, so I definitely support that idea.

**Comment by geoffthemedio** [comment]
Currently stacking groups are implemented so that the first effect to act on an object with a particular stacking group set is the only effect that actually executes on that object. All others with that stacking group don't execute for that object. This means that what effect is applied to an object depends on the order of execution of the effects.

To instead have a priority ranking system, things will get complicated. An arbitrary effect can't be undone / unapplied when another is applied with a higher priority, so there would need to be another layer of caching for each object of all effects for each stacking group that could act on it. Then in a second pass, the top-ranked effect could be applied. But it's not clear how this would interact with effect priorities and the resulting ordering of effects.

It's unnecessary to do this extra processing because it can be written into the scripts in a case-specific and -appropriate way, without so many complications or ambiguities with the general situation.

**Comment by Vezzra** [comment]
Hm, right, the interaction between such a "stacking priority" and the recently added effect priorities is something to consider. OTOH, that means we actually already _have_ a way to prioritize stacking: Just give the effectsgroup you want to take precedence a higher (or lower, whatever makes it to get executed first) effect priority - et voilà, "stacking priority" achieved by effect priority.

And we can still consider adding a "stacking priority" that just works within the same effect priority. That simplifies the implementation, as we only have to gather all effects of a stacking group within the effect priority that is currently executed.

Of course, that still requires the extra layer of caching and the second pass to actually apply the top-ranked effectsgroup. But I still think the scripting of stacking effects would benefit so much from the simplifications made possible by this, that it would be worth this added complexity on the C++ side.

Together with the possibilites the effect priorities already offer, this should cover a wide range of things content scripters might want to do.

Of course such kind of support isn't strictly _necessary_, as it obviously _can_ be done without that (after all, it _is_ already done in a couple of cases), but honestly, these implementations are ackward, clunky, complicated, error prone and a pain to maintain. I really think being able to prioritize the stacking of effects will help a lot on that front.

I _don't_ believe that just adding a possibility to disable auto-generated effects will be sufficient to reduce all the complexity required e.g. to get "best shield" part stacking working right. That will help a lot, but not solve all the problems here.

**Comment by geoffthemedio** [comment]
Removing the auto-generated effects won't make it trivial to get the "best" effect, but would remove the need to have a separate macro to undo the auto-generated effect, which is what I was complaining about.

I don't know what you mean by "a "stacking priority" that just works within the same effect priority"

**Comment by MatGB** [comment]
We've now got, but haven't started using, the ability to assign an order to effects so that the ones we want to happen first do so.

Ergo if we set, say, Sensors as high priority, Neutron Scanner as mid priority and Active Radar as low priority then, in theory, the old stacking group system can still be used and as Sensors go first, that'll be the one you get, all the time.  In theory.

**Comment by geoffthemedio** [comment]
You suggested adding this, after describing what was already there, so I'm not sure what you're trying to say...

**Comment by Vezzra** [comment]
> I don't know what you mean by "a "stacking priority" that just works within the same effect priority"

By that I mean that "stacking priorities" would only be evaluated within the same effect priority level, so you don't have to do the caching and determining of the top-ranked effectsgroup accross several effect priorities, which would require some very complicated forms of resolution.

That is, the caching and determining of the top-ranked effectsgroup for each stacking group is done separately and independently for each effect priority level. This way, the "stacking priority" would be just a means of more fine grained control over which effectsgroup takes precedence when having multiple effectsgroups that have the same effect priority **_and**_ belong to the same stacking group.

**Comment by Vezzra** [comment]
> Ergo if we set, say, Sensors as high priority, Neutron Scanner as mid priority and Active Radar as low priority then, in theory, the old stacking group system can still be used and as Sensors go first, that'll be the one you get, all the time. In theory.

That should work perfectly fine in practice too! :smile:

**Comment by Vezzra** [comment]
> You suggested adding this, after describing what was already there, so I'm not sure what you're trying to say...

I guess he only wants to say the same what I already mentioned above: the effect priorities can already be used to implement "stacking priority", because the current stacking mechanism is based on order of execution, which can be controlled by the effect priority mechanic.

**Comment by geoffthemedio** [comment]
yeah, sorry: forgot that MatGB and Vezzra aren't the same person.

**Pull request merged**